### PR TITLE
[Part4] grpc: enable raw encoding

### PR DESCRIPTION
**Pull request by vivekrsharma**
Adds support for raw encoding over grpc.

### CODE REVIEW

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 1]
```diff
@@ -50,18 +60,44 @@
...
+                .build();
+        byte[] sentMessage = "abc".getBytes(Charsets.UTF_8);
+
+        Metadata rawMetadata = GrpcUtils.setupYarpcHeaders();
+        rawMetadata.put(Metadata.Key.of("Rpc-Encoding", Metadata.ASCII_STRING_MARSHALLER), "raw");
+        EchoRawStub rawBlockingStub =
+            io.grpc.stub.MetadataUtils.attachHeaders(new EchoRawStub(rawchannel), rawMetadata)
+                .withDeadlineAfter(1000, TimeUnit.MILLISECONDS);
+        byte[] gotMessage = rawBlockingStub.echo(sentMessage);
+        if (Arrays.equals(gotMessage, gotMessage)) {
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: You mean `Arrays.equals(gotMessage, sentMessage)`?

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 2]
```diff
@@ -50,18 +60,44 @@
...
+          return CompletableFuture.completedFuture(
+              new CrossdockResponse().success("server said " + Arrays.toString(gotMessage)));
+        } else {
+          return CompletableFuture.completedFuture(
+              new CrossdockResponse()
+                  .error(
+                      "Expected "
+                          + Arrays.toString(sentMessage)
+                          + "but server said "
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: missing a space before "but"

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 3]
```diff
@@ -109,4 +145,69 @@ public Pong echo(Ping request) {
...
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream(BUF_SIZE);
+        byte[] chunk = new byte[BUF_SIZE];
+        int readBytes = 0;
+        do {
+          readBytes = stream.read(chunk);
+          buffer.write(chunk, 0, readBytes);
+        } while (readBytes == chunk.length);
+        return buffer.toByteArray();
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Use slf4j?

Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Or maybe just re-throw as a RuntimeException?

Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: I like idea of RuntimeException.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 4]
```diff
@@ -50,18 +59,44 @@
...
     switch (encoding) {
+      case "raw":
+        ManagedChannel rawchannel =
+            ManagedChannelBuilder.forAddress(request.getParam("server"), 8089)
+                .usePlaintext(true)
+                .build();
+        byte[] sentMessage = "abc".getBytes(Charsets.UTF_8);
+
+        Metadata rawMetadata = GrpcUtils.setupYarpcHeaders();
+        rawMetadata.put(Metadata.Key.of("Rpc-Encoding", Metadata.ASCII_STRING_MARSHALLER), "raw");
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Same comment about making this a constant

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 5]
```diff
@@ -50,18 +59,44 @@
...
         ManagedChannel channel =
             ManagedChannelBuilder.forAddress(request.getParam("server"), 8089)
                 .usePlaintext(true)
                 .build();
         Ping ping = Ping.newBuilder().setBeep("abc").build();
 
-        Metadata metadata = GrpcUtils.setupYarpcHeaders();
-        metadata.put(Metadata.Key.of("Rpc-Encoding", Metadata.ASCII_STRING_MARSHALLER), "json");
+        Metadata jsonMetadata = GrpcUtils.setupYarpcHeaders();
+        jsonMetadata.put(Metadata.Key.of("Rpc-Encoding", Metadata.ASCII_STRING_MARSHALLER), "json");
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: constant

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 6]
```diff
@@ -109,4 +144,68 @@ public Pong echo(Ping request) {
...
+     */
+    @Override
+    public byte[] parse(InputStream stream) {
+
+      try {
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream(BUF_SIZE);
+        byte[] chunk = new byte[BUF_SIZE];
+        int readBytes = 0;
+        do {
+          readBytes = stream.read(chunk);
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: My Java is a little rusty, but can't you use https://docs.oracle.com/javase/7/docs/api/java/io/BufferedInputStream.html?

Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: well, we would need BufferedOutputStream. The constructor also needs output stream which and that would be ByteArrayOutputStream.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import works.crossdock.CrossdockRequest;
+import works.crossdock.CrossdockResponse;
+import works.crossdock.client.Behavior;
+
+@Slf4j
+public class EchoGrpcBehavior implements Behavior {
+  public static final String JSON_ENCODING = "json";
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Isn't this somewhere else too?

Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: well, that's in json package but it is private there. want me to make that public? I'd prefer this way.

**Pull request merged**